finite_difference/src/field_mod.f90

```diff
@@ -1064,14 +1214,55 @@ subroutine halo_exchange(self, depth)
     ! Locals
     integer :: exch  !> Handle for exchange
 
-    ! Synchronize the data if the data is in an external device
-    self%data = self%get_data()
+    ! Make sure the data from each halo is in the host before the communication
+    call self%read_halo_from_device(comm=JPlus, blocking=.false.)
+    call self%read_halo_from_device(comm=Jminus, blocking=.false.)
+    call self%read_halo_from_device(comm=IPlus, blocking=.false.)
+    call self%read_halo_from_device(comm=IMinus, blocking=.true.)
 
     ! Execute the halo exchange
     call exchange_generic(b2=self%data, handle=exch, &
                           comm1=JPlus, comm2=Jminus, comm3=IPlus, comm4=IMinus)
+
+    ! Synchronize the data back to the accelerator device if necessary
+    call self%write_halo_to_device(comm=JPlus, blocking=.false.)
+    call self%write_halo_to_device(comm=Jminus, blocking=.false.)
+    call self%write_halo_to_device(comm=IPlus, blocking=.false.)
+    call self%write_halo_to_device(comm=IMinus, blocking=.true.)
+
   end subroutine halo_exchange
   
+  subroutine read_halo_from_device(self, comm, blocking)
+    use parallel_comms_mod, only: isrcsend, jsrcsend, nxsend, nysend
+    implicit none
+    class(r2d_field), target, intent(inout) :: self
+    integer, intent(in) :: comm
+    logical, intent(in) :: blocking
+
+    ! Get location and dimension to read data for the given communicator
+    if (isrcsend(comm) > 0) then
+        !write(*,*) "Read halo from device: I=", isrcsend(comm), ", J=", &
```

Comment:
Please could you rm this commented code (and again in `write_halo_to_device`).

Reply:
Removed